### ISSUE TITLE
"RangeError: Maximum call stack size exceeded" when writing canonical visual markdown in source mode

### ISSUE THREAD

**Issue by salim-b** [rstudio/rstudio#8116]
### System details

    RStudio Edition : Desktop
    RStudio Version : 1.4.949
    OS Version      : Ubuntu 20.04
    R Version       : 4.0.3

### Steps to reproduce the problem

1. Download and open my test file [`questionnaire.Rmd`](https://gist.github.com/salim-b/ebaa0f659757cbda4b84fdd1b4519569) in RStudio and enable **`Write canonical visual markdown in source mode`** under <kbd>Tools</kbd>→<kbd>Project Options...</kbd>→<kbd>R Markdown</kbd>.

2. Edit the test file (e.g. add a blank line somewhere) and try to save it.

3. Instead of being saved, a popup dialog with the message

   ```
   RangeError: Maximum call stack size exceeded
   ```

   will appear. The test file will stay "unsaved" forever.

### Describe the problem in detail

The test file I provided is rather "complex" (large R code chunks) but in correct Markdown syntax.

Pandoc has no problems parsing it:

```sh
pandoc --from=markdown --to=markdown https://gist.githubusercontent.com/salim-b/ebaa0f659757cbda4b84fdd1b4519569/raw/8496604b53e4296f65e155c7f24d57b9fc5cb38b/questionnaire.Rmd
```

I don't know on what Pandoc version RStudio's Visual Markdown Editor relies. My system's Pandoc version is `2.11.0.2`, the one bundled with RStudio (found under env var `RSTUDIO_PANDOC`) is the same version. If that matters.

### Describe the behavior you expected

I expected the file to be "canonicalized" and successfully saved.

- [x] I have read the guide for [submitting good bug reports](https://github.com/rstudio/rstudio/wiki/Writing-Good-Bug-Reports).
- [x] I have installed the latest version of RStudio, and confirmed that the issue still persists.
- [x] If I am reporting a RStudio crash, I have included a [diagnostics report](https://support.rstudio.com/hc/en-us/articles/200321257-Running-a-Diagnostics-Report).
- [x] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

**Comment by ronblum** [comment]
@salim-b Thank you for raising the issue!

Note: This also happens when trying to switch from source mode to the visual editor.

**Comment by kevinushey** [comment]
Note: also saw these warnings

<img width="406" alt="Screen Shot 2020-10-20 at 10 45 52 AM" src="https://user-images.githubusercontent.com/1976582/96624317-6f496080-12c1-11eb-902e-8e2b910237c3.png">

**Comment by kevinushey** [comment]
Here's the devtools stack trace:

<img width="960" alt="Screen Shot 2020-10-20 at 11 05 11 AM" src="https://user-images.githubusercontent.com/1976582/96626294-2941cc00-12c4-11eb-816d-44638622a79b.png">

@jjallaire do you want to take a look? It looks like the issue here is somehow specific to the very large R chunk used at the start of the document.

**Comment by jjallaire** [comment]
Yes, I'll take a look. In the R Markdown package we actually increase the Pandoc stack size (IIRC to deal with large base64 encodings, but there may be other things this is "making work"). We probably just need to increase the stack size for visual mode invocations as well.

**Comment by jjallaire** [comment]
This issue was actually spreading> 50,000 arguments onto `String.fromCharCode`. Fix is here: https://github.com/rstudio/rstudio/commit/2f7d7c8fe5e68111b6c79c6ad909e4e57ab9a20c.

Marking as required for next preview b/c it affects the ability of users to save Rmd files with large code chunks.

**Comment by salim-b** [comment]
> Fix is here: [2f7d7c8](https://github.com/rstudio/rstudio/commit/2f7d7c8fe5e68111b6c79c6ad909e4e57ab9a20c).

I can confirm this is fixed in daily build 1.4.958. Thank you!
